lib/byron/bench/Restore.hs

```diff
@@ -0,0 +1,468 @@
+    ( runMigrationSilent )
+import Fmt
+    ( fmt, pretty, (+|), (+||), (|+), (||+) )
+import Ouroboros.Network.NodeToClient
+    ( NodeToClientVersionData (..) )
+import Say
+    ( sayErr )
+import System.Environment
+    ( getEnv )
+import System.FilePath
+    ( (</>) )
+import System.IO
+    ( BufferMode (..)
+    , IOMode (..)
+    , hFlush
+    , hSetBuffering
+    , stderr
+    , stdout
+    , withFile
+    )
+import System.IO.Temp
+    ( withSystemTempFile )
+
+import qualified Cardano.BM.Configuration.Model as CM
+import qualified Cardano.BM.Data.BackendKind as CM
+import qualified Cardano.Wallet as W
+import qualified Cardano.Wallet.DB.Sqlite as Sqlite
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+
+-- | Run all available benchmarks. Can accept one argument that is a target
+-- network against which benchmarks below should be ran
+--
+main :: IO ()
+main = do
+    hSetBuffering stdout NoBuffering
+    hSetBuffering stderr NoBuffering
+
+    (_logCfg, tr) <- initBenchmarkLogging Info
+    installSignalHandlers (return ())
+
+    -- TODO: We might want to support testnet via a
+    -- `--benchmark-arguments "testnet"`
+    let network = Mainnet
+
+    -- Reading setup from env vars.
+    --
+    -- If you usually run ./scripts/mainnet.sh in the cardano-node repository,
+    -- it can be useful to set
+    --
+    --     export BYRON_CONFIGS=$CARDANO_NODE_DIR/configuration
+    --     export NODE_DB=$CARDANO_NODE_DIR/db
+    --
+    -- before running this benchmark manually with stack.
+    configs <- getEnv "BYRON_CONFIGS"
+    nodeDB <- getEnv "NODE_DB"
+
+    let genesisHash = "5f20df933584822601f9e3f8c024eb5eb252fe8cefb24d1317dc3d432e940ebb"
+    let socketPath = "/tmp/socket"
+    let args =
+            [ "run"
+            , "--database-path", nodeDB
+            , "--genesis-file", configs </> "mainnet-genesis.json"
+            , "--genesis-hash", genesisHash
+            , "--topology", configs </> "mainnet-topology.json"
+            , "--socket-path", socketPath
+            , "--config", configs </> "configuration-mainnet.yaml"
+            , "--port", "7776"
+            ]
+    -- NOTE: It is warned against using @NoStream@ to supress output on POSIX
+    -- systems, since it could lead to the process throwing errors.
+    -- It appears to work fine stopping the live view of cardano-node,
+    -- however.
+    let cmd = Command "cardano-node" args (return ()) NoStream NoStream
+    void $ withBackendProcess nullTracer cmd $ do
+        case network of
+            Testnet _ -> error "Testnet: not supported yet"
+            Mainnet -> do
+                prepareNode (Proxy @'Mainnet) socketPath
+                runBenchmarks
+                    [ bench ("restore " <> "mainnet" <> " seq")
+                        (bench_restoration @'Mainnet @ShelleyKey
```

Comment:
Doh! ShelleyKey aren't actually a thing on the Byron network. Byron network only uses `ByronKey` and `IcarusKey`! 
Benchmarks should be made for these two! The % one can be limited to only the IcarusKey though.

Reply:
👍 6632d448cfdc1fdedb7353cf7140aa77937d91f0